libbeat/common/mapstr.go

```diff
@@ -86,51 +105,60 @@ func (m MapStr) String() string {
 	return string(bytes)
 }
 
-// UnmarshalYAML helps out with the YAML unmarshalling when the target
-// variable is a MapStr. The default behavior is to unmarshal nested
-// maps to map[interface{}]interface{} values, and such values can't
-// be marshalled as JSON.
+// MergeFields merges the top-level keys and values in each source hash (it does
+// not perform a deep merge). If the same key exists in both, the value in
+// fields takes precedence. If underRoot is true then the contents of the fields
+// MapStr is merged with the value of the 'fields' key in ms.
 //
-// The keys of map[interface{}]interface{} maps will be converted to
-// strings with a %v format string, as will any scalar values that
-// aren't already strings (i.e. numbers and boolean values).
-//
-// Since we want to modify the receiver it needs to be a pointer.
-func (ms *MapStr) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	var result map[interface{}]interface{}
-	err := unmarshal(&result)
-	if err != nil {
-		panic(err)
+// An error is returned if underRoot is true and the value of ms.fields is not a
+// MapStr.
+func MergeFields(ms, fields MapStr, underRoot bool) error {
+	if ms == nil || fields == nil {
+		return nil
+	}
+
+	fieldsMS := ms
+	if !underRoot {
+		f, ok := ms[FieldsKey]
+		if !ok {
+			fieldsMS = make(MapStr, len(fields))
+			ms[FieldsKey] = fieldsMS
+		} else {
+			// Use existing 'fields' value.
+			fieldsMS, ok = f.(MapStr)
+			if !ok {
+				return ErrorFieldsIsNotMapStr
+			}
+		}
 	}
-	*ms = cleanUpInterfaceMap(result)
+
+	// Add fields and override.
```

Comment:
@andrewkroh Seems like the old comment disappeared. The main reason I was thinking about logging is for debugging reason in case we get the info that some fields disappear. Lets leave it out for now and discuss it in case we hit such a case.